game/tools/bigworld/server/message_logger/message_logger.1
.TH MESSAGE_LOGGER 1
.SH NAME
message_logger - BigWorld cluster logging daemon

.SH SYNOPSIS
.B message_logger
[ \fB--daemon\fR ]
[ \fB-u\fR \fIuid\fR ]
[ \fB-p\fR \fIprocess_type\fR ]
[ \fB-d\fR \fIprocess_type\fR ]
[ \fB-l\fR \fIlogger_id\fR ]
[ \fB-v\fR ]
[ \fB--version\fR ]
[ \fB-c|--config\fR \fIfile\fR ]
[ \fB-h|--help\fR ]
[ \fB-t|--text\fR ]
[ \fB-q|--quiet\fR ]
[ \fB-o\fR \fIfile\fR ]
[ \fB-e\fR \fIfile\fR ]
[ \fB+{log level}\fR ]
[ \fB-{log level}\fR ]
[ \fIlog_directory\fR ]

.SH SUMMARY
.B MessageLogger
is a process used for logging output from BigWorld processes
within a cluster. It is a standalone Linux process that can be run on
any machine in a BigWorld cluster that is running BWMachined.

.SH OPTIONS
.SS "Generic Program Options"
.TP
.BI \-c " config"
Specify a config file to use.

If a configuration file is not specified, MessageLogger will attempt to
locate the tools directory in \fI/etc/bigworld.conf\fR under which it will look
for \fImessagelogger/messagelogger.conf\fR.

If \fI/etc/bigworld.conf\fR does not exist, the default is
\fImessage_logger.conf\fR in the current directory.
.TP
.B \-\^\-daemon
Run the MessageLogger process as a daemon.
.TP
.BI \-e " file"
Redirect standard error to \fIfile\fR. This option is useful when
running MessageLogger as a daemon.
.TP
.B \-\^\-help
Print a usage message briefly summarizing these command-line options,
then exit.
.TP
.BI \-o " file"
Redirect standard output to \fIfile\fR. This option is useful when
running MessageLogger as a daemon.
.TP
.BR \-q ", " \-\^\-quiet
Quiet mode. Do not display messages generated by this process. By default these
are output to standard error.
.TP
.BR \-t ", " \-\^\-text
Generate text logs as well as binary logs.
.TP
.B \-v
Enable verbose debugging output.
.TP
.B \-\^\-version
Print the version of message_logger binary, then exit.
.TP
.B log_directory
Output will be stored in this directory. If not specified, this will be read
from the config file \fBlogdir\fR option.
.SS "General Log Filters"
.TP
.BI \-l " logger_id"
Only log messages destined for \fIlogger_id\fR.

Logs for the Tools process type do not have a logger ID associated with them
and will not be filtered out (this allows server start messages to be recorded
and enables log-viewing functionality from the last server start time).
.TP
.BI \-u " uid"
Log only messages from processes run by the user with the specified uid. If
\fIuid\fR is '\fIall\fR' or not specified, messages from all users will be
logged.
.SS "Process Type Log Filter"
.TP
.BI \-d " process_type"
Disable logging of messages from processes with process type name
\fIprocess_type\fR. Multiple names may be specified using multiple \fB\-d\fR
options. Process type inclusion (\fB\-p\fR) and exclusion (\fB\-d\fR) options
should not be used together. By default messages from processes of any name are
logged.
.TP
.BI \-p " process_type"
Enable logging of messages from processes with process type name
\fIprocess_type\fR. Specifying this option disables logging of other process
types, however multiple names may be specified using multiple \fB\-p\fR options.
Process type inclusion (\fB\-p\fR) and exclusion (\fB\-d\fR) options should not
be used together. By default messages from processes of any name are logged.
.TP
Appropriate \fIprocess_type\fR names include:
\fBCellAppMgr\fR,
\fBCellApp\fR,
\fBBaseApp\fR,
\fBBaseAppMgr\fR,
\fBDBMgr\fR,
\fBDBApp\fR,
\fBDBAppMgr\fR,
\fBLoginApp\fR,
\fBServiceApp\fR,
\fBTools\fR.
.SS "Log Level Filter"
.TP
.BI -[TRACE|DEBUG|INFO|NOTICE|WARNING|ERROR|CRITICAL|HACK|ASSET]
Disable logging of a specific log level. Multiple log levels may be excluded by
using multiple instances of this option. Log level exclusion and inclusion
options should not be used together. By default all log levels are logged.
.TP
.BI +[TRACE|DEBUG|INFO|NOTICE|WARNING|ERROR|CRITICAL|HACK|ASSET]
Enable logging of a specific log level. Specifying this option disables logging
of other log levels, however multiple log levels may be included by using
multiple instances of this option. Log level exclusion and inclusion options
should not be used together. By default all log levels are logged.
.SH SIGNALS
.SS SIGHUP
Begin rolling logs. This process may cause the server to slow down. As this is
only part of the rolling and archiving process, it is instead recommended to
use the log rotation functionality of the MessageLogger init.d script:
.IP
\fB/etc/init.d/bw_message_logger logrotate\fR
.SS SIGUSR1
Validate and update the hostnames cache, both on file and on disk. Once the
hostnames cache has been updated it will no longer be possible to search the
logs for outdated hostnames. Activating the hostnames cache validation process
on a very large cluster of machines during a heavy load may impact the
performance of MessageLogger.
.\" .SH FILES
.SH EXAMPLES
.P
Run MessageLogger for all users using a specific config file
\fI/etc/bigworld/message_logger.conf\fR, excluding trace and debug log levels and
logging to \fI/var/log/bigworld/message_logger\fR.
.IP
\fBmessage_logger -u all -c /etc/bigworld/message_logger.conf -TRACE -DEBUG
/var/log/bigworld/message_logger\fR
.P
Run MessageLogger for uid 1234, using a config file in the current
directory and recording only error and critical log levels. The log output
directory will be read from the config file \fBlogdir\fR option.
.IP
\fBmessage_logger -u 1234 -c ./message_logger.conf +ERROR +CRITICAL\fR
.SH "EXIT STATUS"
.B MessageLogger
will exit with a return status of 0 on graceful shutdown or a value of 1 if a fatal
error was encountered during operations.
.SH "SEE ALSO"
.SS "BigWorld Tools"
mltar(1)
.br


game/tools/bigworld/server/install/message_logger.sh
#!/bin/sh

# Redhat service management information
# chkconfig: 2345 90 40
# description: Logs component messages from BigWorld services.
### BEGIN INIT INFO
# Provides:          message_logger
# Required-Start:    $local_fs $network $syslog $time bwmachined2
# Should-Start:
# Required-Stop:     $local_fs $syslog
# Default-Start:     3 4 5
# Default-Stop:      0 1 2 6
# Short-Description: Start BigWorld Message Logger
# Description:       Start BigWorld Message Logger, the service that receives \
#                    and stores logs from BigWorld processes running within a
#                    BigWorld cluster.
### END INIT INFO

if [ -f "/etc/redhat-release" ]; then
        . /etc/init.d/functions
fi
. /etc/init.d/bw_functions


PROCNAME=message_logger
ML_DIR="$BW_TOOLS_DIR/message_logger"
BIN_INSTALL="$BW_TOOLS_BIN/$PROCNAME"
CONFFILE="$ML_DIR/message_logger.conf"
PIDFILE="$BW_TOOLS_PIDDIR/$PROCNAME.pid"
USING_MLDB=1

# Set this to be an email address to activate summaries of logs mailed to that
# address
#LOGROTATE_SUMMARY_MAIL_TO=programmers@example.com

# The summary mail From field.
LOGROTATE_SUMMARY_MAIL_FROM=bwtools@`hostname`

# The SMTP mail host to send through
LOGROTATE_SUMMARY_MAIL_HOST="mail"

# The summary mail subject prefix, defaults to [bwlog-summary]
#LOGROTATE_SUMMARY_MAIL_SUBJECT_PREFIX=

# The summary flags
LOGROTATE_SUMMARY_FLAGS="Psm"

# The summary severities
LOGROTATE_SUMMARY_SEVERITIES="wec"

# The summary period - this defaults to 1 day
#LOGROTATE_SUMMARY_DAYS=1

#Set this for specifying the summary period in hours instead of days
#LOGROTATE_SUMMARY_HOURS=

# Set this to only perform summary on specific user, otherwise all users
#LOGROTATE_SUMMARY_USER=

# Set this to set the minimum count for summaries 
#LOGROTATE_SUMMARY_MINCOUNT=1

# Include message_logger defaults if available
if [ -f /etc/default/$PROCNAME ] ; then
	. /etc/default/$PROCNAME
fi

if [ ! -d "$ML_DIR" ]; then
	echo "ERROR: '$ML_DIR' doesn't appear to exist."
	echo "       Is /etc/bigworld.conf up to date?"
	exit 1
fi

# Read the message logger configuration file and extract the
# log directory / pid file from it.
if [ ! -f "$CONFFILE" ]; then
	echo "ERROR: Unable to find configuration file '$CONFFILE'."
	exit 1
fi

STORAGE_TYPE=`grep storage_type $CONFFILE | sed -r -e "s/$BW_VAL_REGEX/\1/"`
if [ -z $STORAGE_TYPE ]; then
	echo "ERROR: No storage_type entry in $CONFFILE."
	exit 1
fi

if [ "$STORAGE_TYPE" != "mldb" ]; then
	USING_MLDB=0
fi

# settings for MLDB only
if [ $USING_MLDB -eq 1 ]; then
	ML_LOGDIR=`grep logdir $CONFFILE | sed -r -e "s/$BW_VAL_REGEX/\1/"`
	if [ ! -d $ML_LOGDIR ]; then
		echo "ERROR: '$ML_LOGDIR' does not appear to be a directory."
		exit 1
	fi
	
	# Message Logger lock file for MLDB
	LOCKFILE="$ML_LOGDIR/pid_lock"

	ML_DEFAULT_ARCHIVE=`grep default_archive $CONFFILE | sed -r -e "s/$BW_VAL_REGEX/\1/"`
	if [ -z $ML_DEFAULT_ARCHIVE ]; then
		echo "ERROR: No 'default_archive' specified in '$CONFFILE'."
		exit 1
	fi
fi

# Notify Message Logger to roll its log by sending SIGHUP to it
roll_log() {
	read pid < $PIDFILE
	if [ $? != 0 ]; then
		echo "Unable to read PID from $PIDFILE."
		return 1
	fi

	kill -HUP $pid 2> /dev/null

	if [ $? != 0 ]; then
		echo "Unable to send SIGHUP to $pid"
		return 1
	fi

	return 0
}


case "$1" in

  start)
	# Check for an existing service running
	bw_is_running "$PIDFILE" "message_logger --daemon" "$BW_TOOLS_USERNAME"
	if [ $? != 0 ]; then
		if [ -f "/etc/debian-version" ]; then
			echo -n "already running, rolling logs. "
		else
			echo "already running, rolling logs. "
		fi
		
		roll_log
		
		if [ $? != 0 ]; then
			if [ -f "/etc/debian-version" ]; then
				bw_print_failure
			fi
			echo "Failed to roll log."
			exit 1
		fi

		if [ -f "/etc/debian-version" ]; then
			bw_print_success
		fi
		exit 0
	elif [ $USING_MLDB -eq 1 ]; then
		# If no message_logger service is running, remove the orphaned lock file
		bw_is_running "$LOCKFILE" "message_logger --daemon" "$BW_TOOLS_USERNAME"
	fi

	LOGFILE="$BW_TOOLS_LOG_DIR/message_logger.log"
	if [ -z "$LOGGER_ID" ]; then
		OPTIONS=" -o \"$LOGFILE\"  -e \"$LOGFILE\" -c \"$CONFFILE\" --pid \"$PIDFILE\"" 
	else
		OPTIONS=" -o \"$LOGFILE\"  -e \"$LOGFILE\" -l \"$LOGGER_ID\"  -c \"$CONFFILE\" --pid \"$PIDFILE\"" 
	fi

	if [ -f "/etc/redhat-release" ]; then
		echo -n "Starting bw_message_logger: "
		daemon su " -c \"$BIN_INSTALL --daemon \
			$OPTIONS \" $BW_TOOLS_USERNAME"
		echo
	else
		su -c "$BIN_INSTALL --daemon \
			$OPTIONS" $BW_TOOLS_USERNAME
	fi

	RETVAL=$?
	if [ $RETVAL != 0 ]; then
		if [ -f "/etc/debian-version" ]; then
			bw_print_failure
		fi
		exit 1
	else

		bw_has_started "$PIDFILE" "message_logger --daemon" "$BW_TOOLS_USERNAME"
		if [ $? == 0 ]; then
			if [ -f "/etc/debian-version" ]; then
				bw_print_failure
			fi
			echo "message_logger doesn't appear to have started."
			echo "Check '$LOGFILE' for errors"
			exit 1
		fi
	fi

	if [ -f "/etc/debian-version" ]; then
		bw_print_success
	fi
	RETVAL=0
	;;


  stop)

	bw_stop_process "$PROCNAME" \
					"$PIDFILE" \
					"message_logger --daemon" \
					"$BW_TOOLS_USERNAME"
	RETVAL=$?
	;;

  summarise)
  	RETVAL=0

	# Currently, only MLDB supports summarise function
  	if [ $USING_MLDB -eq 0 ]; then
  		RETVAL=0
		break
  	fi

	cd "$ML_DIR"
	if [[ -z $LOGROTATE_SUMMARY_MAIL_TO ]]; then
		# not configured
		RETVAL=0
		break
	else
		# do all the compulsory configured options
		FLAGS="--mail-to=$LOGROTATE_SUMMARY_MAIL_TO \
			--mail-from=$LOGROTATE_SUMMARY_MAIL_FROM \
			--mail-host=$LOGROTATE_SUMMARY_MAIL_HOST \
			--summary-flags=$LOGROTATE_SUMMARY_FLAGS \
			--severities=$LOGROTATE_SUMMARY_SEVERITIES"

		if [[ ! -z "$LOGROTATE_SUMMARY_MAIL_SUBJECT_PREFIX" ]]; then
			FLAGS="$FLAGS --mail-summary=\"$LOGROTATE_SUMMARY_MAIL_SUBJECT_PREFIX\""
		fi


		if [[ ! -z $LOGROTATE_SUMMARY_HOURS ]]; then
			FLAGS="$FLAGS --hours $LOGROTATE_SUMMARY_HOURS"
		else
			if [[ ! -z $LOGROTATE_SUMMARY_DAYS ]]; then
				FLAGS="$FLAGS -- days $LOGROTATE_SUMMARY_DAYS"
			fi
		fi
		
		if [[ -z $LOGROTATE_SUMMARY_USER ]]; then
			FLAGS="$FLAGS --all-users"
		else
			FLAGS="$FLAGS -u $LOGROTATE_SUMMARY_USER"
		fi

		if [[ ! -z $LOGROTATE_SUMMARY_MINCOUNT ]]; then
			FLAGS="$FLAGS --summary-min=$LOGROTATE_SUMMARY_MINCOUNT"
		fi

		echo -n "Summarising logs: "
		if python mlsum.py $FLAGS; then
			bw_print_success
		else
			bw_print_failure
		fi
	fi
  	;;

  logrotate)
	RETVAL=0

	# If the backend storage is not MLDB, just roll log and exit
	if [ $USING_MLDB -eq 0 ]; then
  		roll_log  		
  		RETVAL=$?

  		if [ RETVAL -eq 0 ]; then
  			echo "Successfully rolled log."
  		else
  			echo "Failed to roll log."
  		fi
		
		# log rotation is done
		break
  	fi

	echo -n "Rolling message_logger logs: "
	cd "$ML_DIR"

	# Send a SIGHUP to message_logger
	python mlroll.py "$ML_LOGDIR"
	if [ $? != 0 ]; then
		echo "Failed to mlroll '$ML_LOGDIR'."
		RETVAL=$[ $RETVAL + 1 ]
	fi

	# Archive all the old logs
	nice -n 19 \
		python mltar.py -cvr --all-users --hours-old 23 \
			-f "$ML_LOGDIR"/message_logs.tar.gz "$ML_LOGDIR"
	if [ $? != 0 ]; then
		echo "Failed to mltar '$ML_LOGDIR'."
		RETVAL=$[ $RETVAL + 1 ]
	fi

	# The logrotate step will most often be run as root, so chown the
	# newly created archive back to the user that owns the rest of the logs.
	if [ -f $ML_DEFAULT_ARCHIVE ]; then
		chown "$BW_TOOLS_USERNAME:$BW_TOOLS_GROUPNAME" "$ML_DEFAULT_ARCHIVE"
		if [ $? != 0 ]; then
			echo "Failed to chown '$ML_DEFAULT_ARCHIVE' to user '$BW_TOOLS_USERNAME'"
			RETVAL=$[ $RETVAL + 1 ]
		fi
	fi

	# Remove any unpacked / 'dirty' logs
	python mlrm.py --dirty "$ML_LOGDIR"
	if [ $? != 0 ]; then
		echo "Failed to remove dirty files in '$ML_LOGDIR'."
		RETVAL=$[ $RETVAL + 1 ]
	fi

	# Do a regular mlrm pass over the logdir to blow away any empty user dirs
	python mlrm.py -a --hours 24 "$ML_LOGDIR"
	if [ $? != 0 ]; then
		echo "Failed to cleanup empty user dirs in '$ML_LOGDIR'."
		RETVAL=$[ $RETVAL + 1 ]
	fi

	if [ $RETVAL != 0 ]; then
		bw_print_failure
		RETVAL=1
	fi
	;;

  # Summarise log file.
  summarise_log)

	echo -n "Summarising WebConsole log file: " 
	SUMMARISE_OPTIONS=" "
	if [ ! -z $BW_TOOLS_SUMMARY_MAIL_FROM ]; then
		if [ ! -z $BW_TOOLS_SUMMARY_MAIL_TO ]; then
			SUMMARISE_OPTIONS="$SUMMARISE_OPTIONS \
				--mail-to=$BW_TOOLS_SUMMARY_MAIL_TO \
				--mail-from=$BW_TOOLS_SUMMARY_MAIL_FROM"
		fi
	fi

	SUMMARY_CMD="./summarise_logs.py$SUMMARISE_OPTIONS \
		message_logger"

	su - $BW_TOOLS_USERNAME -c "cd $BW_TOOLS_DIR/misc; $SUMMARY_CMD"	
	
	if [ $? != 0 ]; then
		bw_print_failure
		RETVAL=1
	else
		bw_print_success
		RETVAL=0
	fi
	echo
  ;;

  status)

	echo -n "Status of $PROCNAME: "
	if [[ ! -e $PIDFILE ]]; then
		echo "stopped"
		RETVAL=1
	else
		# Extract the PID from the pid file
		PID=`head -n 1 $PIDFILE 2>/dev/null`
		if [ $? != 0 ]; then
			echo "Unable to read pid from '$PIDFILE'."
			RETVAL=1
		else

			# Check if the pid is in the process list
			ps -p $PID > /dev/null 2>&1
			if [ $? == 0 ]; then
				echo "running"
				RETVAL=0
			else
				echo "pid file exists, but no process running as `head -n 1 $PIDFILE`"
				RETVAL=1
			fi
		fi
	fi
	;;


  restart)

	$0 stop
	RETVAL=$?
	$0 start
	RETVAL=$[ $RETVAL + $? ]
	;;

  version)

	$BIN_INSTALL --version
	;;

  *)
	echo "Usage: $PROCNAME {start|stop|status|restart}"
	exit 1
	;;
esac

exit $RETVAL
